virtual_rainforest/core/constants_loader.py

```diff
@@ -0,0 +1,53 @@
+"""The :mod:`~virtual_rainforest.core.constants` module is used to store constants that
+are used across the Virtual Rainforest. This includes universal constants, such as the
+strength of gravity, but also model constants that are shared between multiple models,
+such as the depth of the biogeochemically active soil layer.
+
+At the moment, no constants are actually stored in this module. It currently only
+contains the :attr:`CONSTANTS_REGISTRY`, which all model constants classes should be
+registered in. This allows for all model constants to be documented neatly.
+"""  # noqa: D205, D415
+
+from typing import Any
+
+from virtual_rainforest.core.config import Config
+from virtual_rainforest.core.exceptions import ConfigurationError
+from virtual_rainforest.core.logger import LOGGER
+from virtual_rainforest.core.registry import MODULE_REGISTRY
+
+
+def load_constants(config: Config, model_name: str, class_name: str) -> Any:
+    """Load the specified constants class.
+
+    Any constants that are supplied for this class in the config are used to populate
+    the class, for all other constants default values are used.
+
+    Args:
+        config: A validated Virtual Rainforest model configuration object.
+        model_name: Name of the model the constants belong to
+        class_name: Name of the specific dataclass the constants belong to
+
+    Returns:
+        A constants class populated using the information provided in the config
+    """
+
+    # Extract dataclass of interest from registry
+    try:
+        constants_class = MODULE_REGISTRY[model_name].constants_classes[class_name]
+    except KeyError as excep:
+        LOGGER.critical(f"Unknown constants class: {model_name}.{class_name}")
+        raise excep
+
+    # If the configuration supplies a section for this constant class, use the
+    # ConstantsDataclass.from_config method to check that the arguments are valid for
+    # the specific dataclass, otherwise return the default.
+    try:
+        constants_config = config[model_name]["constants"][class_name]
+        return constants_class.from_config(constants_config)
+    except (KeyError, IndexError):
+        # No constants_config section, so return default instance.
+        return constants_class()  # type: ignore [operator]  # FIXME !!!!!
+    except ConfigurationError:
+        # The from_config call failed, so log and exit.
+        LOGGER.critical("Could not initialise {model_name}.{class_name} from config.")
+        raise
```

Comment:
I'm not sure why `mypy` complains, to be honest, but all in all, this does not look like the correct approach.

Exceptions should be used to deal with unexpected circumstances. Not having constants is not an unexpected circumstance, but a perfectly reasonable use case, so an Exception is not the right control structure here. 

In summary: if the configuration provides a section for this class, great! Use it. If not, just pass an empty dict:
```suggestion
    constants_config = config[model_name]["constants"][class_name] if "constants" in config[model_name] and class_name in config[model_name]["constants"] else {}
    try:
        return constants_class.from_config(constants_config)
    except ConfigurationError:
        # The from_config call failed, so log and exit.
        LOGGER.critical("Could not initialise {model_name}.{class_name} from config.")
        raise
```
As far as I can tell, `from_config" is absolutely happy ingesting an empty dict and there is no meaningful speed gain calling directly the constructor.

Reply:
I'd disagree a bit that this is the wrong approach - that 'easier to ask forgiveness than permission' style is something commonly used in Python. It isn't a style I particularly like, but it is more compact than the long chain of iterative dictionary key checks. But yes - using the empty dictionary should sidestep that `mypy` issue.

Reply:
Well, if the check were really complicated or deep within the function, I would certainly agree, but the meaning of Exception itself suggest that using them for flow control when you are in normal operation is plainly wrong and can lead to confusion and masking other errors that have nothing to do with what you are trying to catch.

Reply:
Moreover, in this particular case, you do need to add those comments explaining what's going on or otherwise there's no obvious reason of why you're catching those exceptions. The code is way more explicit and needs no explanation if you check for the existence of the relevant constants dict in the config, as I suggest.

Reply:
But many examples of "idiomatic" python use exactly this kind of structure. And the Python glossary specifically mentions this kind of construct (emphasis mine):

> EAFP
> Easier to ask for forgiveness than permission. This common Python coding style **assumes the existence of valid keys or attributes and catches exceptions if the assumption proves false**. This clean and fast style is characterized by the presence of many [try](https://docs.python.org/3/reference/compound_stmts.html#try) and [except](https://docs.python.org/3/reference/compound_stmts.html#except) statements. The technique contrasts with the [LBYL](https://docs.python.org/3/glossary.html#term-LBYL) style common to many other languages such as C.

https://docs.python.org/3/glossary.html#term-EAFP 

I'm not sure how this example differs from that description. Again - it isn't a style I use widely but I do want to understand why what looks to me like a stylistic choice is problematic.

Reply:
> Exceptions should be used to deal with unexpected circumstances. Not having constants is not an unexpected circumstance, but a perfectly reasonable use case, so an Exception is not the right control structure here.

> Exceptions are about exceptional circumstances and things that should not be happening, normally.

I'm struggling with this. 

There are questions of style - I think this implementation is easier to read and maintain than the multiple clause `if` statement - but those don't really matter. But I *do* want to know if I'm missing something subtle about this approach - it seems really useful in some cases. I don't want to be misapplying it and your advice is clearly that I am.

Even in the standard libraries, `try`-ing something that will routinely fail in an expected way and then trapping those failures with `except` seems widespread. An example is `os.exists`, where a file not existing will happen all the time.

https://github.com/python/cpython/blob/1de9406f9136e3952b849487f0151be3c669a3ea/Lib/genericpath.py#L16-L22

What am I missing?

Reply:
There might be cases where checking if something works or not is not straight forward, might require way too much code to fully check if it works or the multiple cases in which it might fail. Or simply there is no other way to do the checks (like in the example you mention - if you just check if the file actually exists, you will get the wrong result as the file exist in the case of broken symbolic links; they are just not connected to anything). In those cases, it is way simpler to just use `try...except...`. The conciseness and simplicity of the code clearly outweighs any benefit related to explicitness of the checks. I'm happy to use Exceptions in those cases.

But I don't think that having that possibility is an excuse when a simple conditional statement can do the job perfectly fine. People abuse of the EAFP idea massively and applies it everywhere, diminishing the actual role of Exceptions which is dealing with expectational circumstances. And they also apply it wrong, as it is your case, as there are two possible sources or errors, the dictionary and the call to `from_config`, with the issues I explained in my previous message.

Reply:
On a different note - on balance I'm going to stick with the `# type: ignore` here. Using `constants_class()` _clearly_ should not throw an error and it _is_ callable. I'd rather ignore it and leave an indication that `mypy` is being a bother than do something unnecessary just to keep `mypy` quiet.